File: web/src/lib/messages/leave.ts
import {
  bufferToHeader,
  bufferToString,
  HEADER_SIZE,
  headerToBuffer,
  MessageType,
  type Message,
} from ".";

export default class LeaveMessage implements Message {
  id: number;
  sentAt: bigint;
  playerId: string;

  constructor(id: number, sentAt: bigint, playerId: string) {
    this.id = id;
    this.sentAt = sentAt;
    this.playerId = playerId;
  }

  static fromBuffer(buffer: Uint8Array): LeaveMessage {
    const header = bufferToHeader(buffer);

    const playerIdOffset = HEADER_SIZE;
    const { value: playerId } = bufferToString(buffer, playerIdOffset);

    return new LeaveMessage(header.id, header.sentAt, playerId);
  }

  toBuffer(): Uint8Array {
    const encodedHeader = headerToBuffer(this);

    const encoder = new TextEncoder();
    const encodedPlayerId = encoder.encode(this.playerId);

    const buffer = new Uint8Array(HEADER_SIZE + 1 + encodedPlayerId.byteLength);
    buffer[0] = MessageType.Leave;
    buffer.set(encodedHeader, 1);
    buffer[HEADER_SIZE] = encodedPlayerId.byteLength;
    buffer.set(encodedPlayerId, HEADER_SIZE + 1);

    return buffer;
  }
}

File: internal/messages/leave.go
package messages

type Leave struct {
	Header
	PlayerID string
}

func DecodeLeave(b []byte) (*Leave, error) {
	header, err := DecodeHeader(b[1:HeaderSize])
	if err != nil {
		return nil, err
	}

	if len(b) < HeaderSize+1 {
		return nil, ErrInvalidLeave
	}

	playerID, err := bufferToString(b, HeaderSize)
	if err != nil {
		return nil, err
	}

	return &Leave{
		Header:   header,
		PlayerID: playerID,
	}, nil
}

func (l *Leave) Type() MessageType {
	return MessageTypeLeave
}

func (l *Leave) Encode() []byte {
	var b []byte
	b = append(b, byte(MessageTypeLeave))
	b = append(b, l.Header.Encode()...)
	b = append(b, uint8(len(l.PlayerID)))
	b = append(b, []byte(l.PlayerID)...)

	return b
}
